### PR TITLE
Fix #9636: report unused type parameter error on the last declaration

### LINKED ISSUE #9636: noUnusedLocals checks for type parameters of merged interface should also check merge targets
**TypeScript Version:**  master

**Code**

```ts
interface C<T> {
}
class C<T> {
  public c: T;
}
```

**Expected behavior:**

pass

**Actual behavior:** 

```
$ node built/local/tsc.js --noUnusedLocals index.ts
index.ts(1,13): error TS6133: 'T' is declared but never used.
```

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes https://github.com/Microsoft/TypeScript/issues/9636

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -14612,8 +14612,15 @@ namespace ts {
         function checkUnusedTypeParameters(node: ClassDeclaration | ClassExpression | FunctionDeclaration | MethodDeclaration | FunctionExpression | ArrowFunction | ConstructorDeclaration | SignatureDeclaration | InterfaceDeclaration) {
             if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
                 if (node.typeParameters) {
+                    // Only report errors on the last declaration for the type parameter container;
+                    // this ensures that all uses have been accounted for.
+                    const symbol = node.symbol && getMergedSymbol(node.symbol);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Can you use `getSymbolOfNode(node)` here instead? `getMergedSymbol` already checks that `node.symbol` is defined.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -16122,7 +16129,7 @@ namespace ts {
 
             if (produceDiagnostics) {
                 checkTypeForDuplicateIndexSignatures(node);
-                checkUnusedTypeParameters(node);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Should anything use checkUnusedTypeParameters except in a deferred context? It looks like it's still called from a bunch of places.

Comment (mhegazy) [inline src/compiler/checker.ts]: signatures do. they do not merge, so after looking at the signature, there is no need to wait, no more uses will show up after that.

File: src/compiler/checker.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: done.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13277,7 +13277,7 @@ namespace ts {
                         checkAsyncFunctionReturnType(<FunctionLikeDeclaration>node);
                     }
                 }
-                if (!(<FunctionDeclaration>node).body) {
+                if (noUnusedIdentifiers && !(<FunctionDeclaration>node).body) {
                     checkUnusedTypeParameters(node);
```
Comment (yuit) [inline src/compiler/checker.ts]: Question: Why you need to do this here? since you will register the node in this [line](https://github.com/Microsoft/TypeScript/pull/9784/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR16132) already

**Comment by sandersn** [comment]
:+1:

File: src/compiler/checker.ts (hunk 3 above)
Comment (yuit) [inline src/compiler/checker.ts]: Talk offline with @mhegazy; this question is actually same as @sandersn

**Comment by yuit** [comment]
:+1:

File: src/compiler/types.ts [hunk 4]
```diff
@@ -2908,7 +2908,6 @@ namespace ts {
         getCancellationToken?(): CancellationToken;
         getDefaultLibFileName(options: CompilerOptions): string;
         getDefaultLibLocation?(): string;
-        getDefaultTypeDirectiveNames?(rootPath: string): string[];
```
Comment (vladima) [inline src/compiler/types.ts]: is this removal intentional?

**Comment by vladima** [comment]
:+1:
